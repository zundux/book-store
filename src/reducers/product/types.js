export const FETCH_PRODUCT_REQUEST = "product/FETCH_PRODUCT_REQUEST";
export const FETCH_PRODUCT_SUCCESS = "product/FETCH_PRODUCT_SUCCESS";
export const FETCH_PRODUCT_FAIL = "product/FETCH_PRODUCT_FAIL";