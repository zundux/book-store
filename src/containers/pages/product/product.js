import React, { Component } from "react";
import { bindActionCreators } from "redux";
import { connect } from "react-redux";
import * as PropTypes from "prop-types";
import { addToCart } from "../../../reducers/cart";
import { fetchProduct } from "../../../reducers/product";
import ProductDetails from "../../../components/product-details";
import LoadingLabel from "../../../components/loading-label";
import ErrorLabel from "../../../components/error-label";
import { Link } from "react-router-dom";

export class Product extends Component {
  componentDidMount() {
    this.props.onFetchProduct(this.props.match.params.id);
  }

  render() {
    let { product, isFetching, errorMessage, onAddToCart } = this.props;

    return (
      <div>
        {!isFetching && !errorMessage && product && (
          <ProductDetails {...{
            product,
            onAddToCart
          }}/>
        )}
        {!isFetching && errorMessage && (
          <ErrorLabel message={`Product not found. ${errorMessage}`}>
            <Link to="/">Back to catalog</Link>
          </ErrorLabel>
        )}
        {isFetching && (
          <LoadingLabel message="Product loading, please wait"/>
        )}
      </div>
    );
  }
}

Product.propTypes = {
  match: PropTypes.object,
  product: PropTypes.object,
  isFetching: PropTypes.bool,
  errorMessage: PropTypes.string,
  onAddToCart: PropTypes.func,
  onFetchProduct: PropTypes.func
};

const mapStateToProps = ({ product }) => ({
  product: product.product,
  isFetching: product.isFetching,
  errorMessage: product.errorMessage
});

const mapDispatchToProps = dispatch => bindActionCreators({
  onAddToCart: addToCart,
  onFetchProduct: fetchProduct
}, dispatch);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Product);
