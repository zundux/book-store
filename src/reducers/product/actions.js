import { FETCH_PRODUCT_FAIL, FETCH_PRODUCT_REQUEST, FETCH_PRODUCT_SUCCESS } from "./types";

export const bookResourceUrl = "https://www.googleapis.com/books/v1/volumes";

export const requestProduct = id => ({
  type: FETCH_PRODUCT_REQUEST,
  id
});

export const receiveProductSuccess = json => ({
  type: FETCH_PRODUCT_SUCCESS,
  product: json
});

export const receiveProductFail = error => ({
  type: FETCH_PRODUCT_FAIL,
  errorMessage: `Could not fetch product. ${error.message}`
});

export const fetchProduct = id => {
  return dispatch => {

    dispatch(requestProduct(id));

    const fetchProductUrl = `${bookResourceUrl}/${id}`;

    return fetch(fetchProductUrl)
      .then(response => response.json())
      .then(json => dispatch(receiveProductSuccess(json)))
      .catch(error => dispatch(receiveProductFail(error)));
  };
};
