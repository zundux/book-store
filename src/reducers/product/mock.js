export const fetchProductData = ({ id: 1, title: "book 1" });
export const fetchProductData2 = ({ id: 2, title: "book 2" });
export const saveProductData = { title: "book 1", amount: 1 };
export const errorMessage = "some error";

export default {
  fetchProductData,
  fetchProductData2,
  saveProductData,
  errorMessage
};