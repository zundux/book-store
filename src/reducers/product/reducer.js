import { FETCH_PRODUCT_FAIL, FETCH_PRODUCT_REQUEST, FETCH_PRODUCT_SUCCESS } from "./types";
import * as R from "ramda";

export const initialStateProduct = {
  product: null,

  id: null,
  isFetching: false,
  errorMessage: null
};

export default (state = initialStateProduct, action) => {
  switch (action.type) {
    case FETCH_PRODUCT_REQUEST:
      return {
        ...state,
        id: action.id,
        isFetching: true,
        errorMessage: null
      };

    case FETCH_PRODUCT_SUCCESS:
      return R.mergeDeepRight(state, {
        product: action.product,
        id: null,
        isFetching: false,
        errorMessage: null
      });

    case FETCH_PRODUCT_FAIL:
      return R.mergeDeepRight(state, {
        isFetching: false,
        errorMessage: action.errorMessage
      });

    default:
      return state;
  }
}