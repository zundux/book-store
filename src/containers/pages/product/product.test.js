import React from "react";
import { Product } from "./product";

describe("Product page", () => {

  it("calls loadProduct when the site has been loaded at first time " +
    "and show loading label", () => {
    const id = 11;
    const match = { params: { id } };
    const product = null;
    const isFetching = true;
    const errorMessage = null;
    const onAddToCart = () => {
    };
    const onFetchProduct = jest.fn();

    const component = shallow(
      <Product {...{
        match,
        product,
        isFetching,
        errorMessage,
        onAddToCart,
        onFetchProduct
      }}/>
    );
    matchSnapshot(component);

    expect(onFetchProduct).toHaveBeenCalled();
    expect(onFetchProduct).toHaveBeenCalledWith(id);
  });

  it("should show product by default", () => {
    const match = { params: { id: 11 } };
    const product = { id: 11 };
    const isFetching = false;
    const errorMessage = null;
    const onAddToCart = () => {
    };
    const onFetchProduct = jest.fn();

    const component = shallow(
      <Product {...{
        match,
        product,
        isFetching,
        errorMessage,
        onAddToCart,
        onFetchProduct
      }}/>
    );
    matchSnapshot(component);
    // todo check page header
    // todo check page title
  });

  it("should show error message if the product not found", () => {
    const match = { params: { id: null } };
    const product = null;
    const isFetching = false;
    const errorMessage = "some error";
    const onAddToCart = () => {
    };
    const onFetchProduct = jest.fn();

    const component = shallow(
      <Product {...{
        match,
        product,
        isFetching,
        errorMessage,
        onAddToCart,
        onFetchProduct
      }}/>
    );
    matchSnapshot(component);
    // todo check page header
    // todo check page title
  });
});