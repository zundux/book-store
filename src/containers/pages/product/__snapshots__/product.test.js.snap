// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product page calls loadProduct when the site has been loaded at first time and show loading label 1`] = `
<div>
  <LoadingLabel
    message="Product loading, please wait"
  />
</div>
`;

exports[`Product page should show error message if the product not found 1`] = `
<div>
  <ErrorLabel
    message="Product not found. some error"
  >
    <Link
      replace={false}
      to="/"
    >
      Back to catalog
    </Link>
  </ErrorLabel>
</div>
`;

exports[`Product page should show product by default 1`] = `
<div>
  <ProductDetails
    onAddToCart={[Function]}
    product={
      Object {
        "id": 11,
      }
    }
  />
</div>
`;
